### PR TITLE
Minor Fixes and Validation

### PR DESCRIPTION
**Pull request by RFBomb**
This PR is a subset of the changes proposed in #161.
I've contained it to what should be the safest changes to implement on the core library.

Some of these changes were designed to facilitate custom IRoboCommand implementations, so I left those in.
The major changes to this are primarily publicizing / adding constructors to various objects to allow for either external unit testing or custom implementations.

I also have a replacement for the ObservableList<T>, which is used by RoboQueue and RoboCopyResultsList.
Now, I'm not saying I regret RoboQueue, because I learned A LOT doing it, but I do think that writing ObservableList was an err on my part. I wound up writing a 'ConcurrentList' which has immutable collections to play nicer with UI updates when multiple results are updated at same time. This is NOT part of this change list, as I think it should be isolated for review.

Linking #177  as it will be resolved in this PR.

The only potentially breaking change here was moving the enums out of SelectionOptions and CopyOptions into the default namespace, which makes it overall easier to use long term. (My bad for putting it in there initially)

### CODE REVIEW

**Comment by RFBomb** [comment]
Rebased after #180

**Comment by RFBomb** [comment]
I want to say I believe this is mostly finalized. I'm starting work on the Extensions namespace shortly, which will include a custom implementation I created to more efficiently MOVE files that reside on the same root drive. ( using File.Move() is significantly quicker than robocopy if on same drive)

**Comment by PCAssistSoftware** [comment]
All appears okay to me, happy to squash and merge if/when you are ready?

**Comment by RFBomb** [comment]
@PCAssistSoftware Thanks for testing all that so quick!
I didn't have internet for a bit and got some more relatively small changes put in.
- Some unit testing for CopyOptions and SelectionOptions methods.
- Minor XML Doc fixes, cleaning up some more compiler warnings/messages.

BIG CHANGE: commit ( https://github.com/tjscience/RoboSharp/pull/179/commits/acbddeb9910bc42c5a277a523260dd7d4a6d6cc9 ) needs some verification on your part, as I know you actually use this functionality and I have nothing to test against. 
Basically, that reworks the authentication to expose it to custom implementations. The actual functionality should be unchanged, but its been moved to a new 'Authentication' static class.  ( This also allows to potentially to authenticate source and destination separate from each other in custom commands ).

I know it didn't pass the current check, I'm investigating that now.

**Comment by PCAssistSoftware** [comment]
@RFBomb - no problem at all, having a quiet day workwise so thought I would devote some time to this.

re big change / ( [acbddeb ](https://github.com/tjscience/RoboSharp/commit/acbddeb9910bc42c5a277a523260dd7d4a6d6cc9)) - actually I don't use the authentication, I recently merged in some changes by @gabrielgt relating to authentication support for .NET 6.0 and had to specifically setup a test machine in a domain to make sure that worked, but that was first time I had ever even passed a username and password to RoboCommand StartAsync.   I can resurrect that test environment and test again with this PR for you.

**Comment by PCAssistSoftware** [comment]
Can't build current PR, got error on line 102 of authentication.cs - user of unassigned local variable 'result'  - that might be why it wasn't passing current check?

**Comment by RFBomb** [comment]
So, the check fails were also present on my computer, because I couldn't build the unit test framework for some reason (even before I added a test). Not really sure what that was about, but I'm updating the project file for the unit tests to the modern Project-SDK style, which resolves it. 

As far as that line 102, I'm really confused by that, as line 102 is the return statement. I cannot build on Net6.0 on my machine (stuck on VS2019), but I believe I know the issue and will have a fix up mometarily. Still investigating the unit tests failing though.

**Comment by PCAssistSoftware** [comment]
Have tested authentication as much as I can by passing domain, username and password to RoboCommand StartAsync and it appears to still work as before.  Passing invalid credentials fails as expected with Win32Exception of 'The user name or password is incorrect' being thrown in ImpersonatedUser.cs and passing correct credentials results in filecopy completing successfully, so as far as I can see all okay in relation to that point.

**Comment by RFBomb** [comment]
Thanks for testing all that! I was confident it would still work, but needed it tested with the updated static class. Consumers should see no difference in it. 

I have just resolved all the issues the unit new tests brought to light. in this latest commit as well. ( I'm really glad those other unit tests were written to validate everything before, as its really brought my confidence level up in these changes)

**Comment by PCAssistSoftware** [comment]
no problem at all - would you like me to merge it now?

**Comment by RFBomb** [comment]
If you are happy with it, I'm happy with it. 
I'm fairly confident this covers all the bases I needed to provide the Extensions package its functionality.
I'm excited to bring over my 'RoboMover' object I created as well, as its significantly quicker than RoboCopy when performing a MOVE action on the same drive ( ex:    F:\Source -> F:\1\2\3\Destination )

**Comment by PCAssistSoftware** [comment]
consider it done :)

looking forward to seeing the Extensions package
